### ISSUE TITLE
Feature Request: Images based on other Base OSes

### ISSUE THREAD

**Issue by CpuID** [rubinius/docker#4]
Would it be possible to get images built off a couple of other Base OS'es at some point?

- Ubuntu 16.04
  - Direct upgrade path from current 14.04
- Alpine Linux
  - Hopefully for image size reasons, I have started building up my own but it would be nice to have an official variant

Thanks!

**Comment by brixen** [comment]
@CpuID yes, I'd love to do this. Here's the problem right now: 1. we build the binaries that we use to build the Docker images in Travis CI; 2. we are limited to the distros that Travis CI has available; and 3. we do this because we were trying to keep the Docker image size small (ie by not including all the stuff needed to build Rubinius).

Initially, we got the Rubinius image down to 95mb, but then there were no C/C++ compilers installed and no C-ext gems could be installed. After we added the C/C++ compilers back, the image was back to ~200mb. So maybe this image size isn't a huge problem.

Perhaps I could set up a multilevel Docker build (similar to the multilevel Travis CI build I have now), where we first build Rubinius on other base images using a build image, then take the binary from that and build more images that only use the binary. Alternatively, maybe I should just finally set up Concourse CI to build everything.

Thoughts?

**Comment by derikson** [comment]
@brixen [Docker 17.06 (stable) is expected](https://github.com/docker/hub-feedback/issues/1039#issuecomment-300291508) to have the [multi-stage build](https://docs.docker.com/engine/userguide/eng-image/multistage-build/) feature, which would allow you to build the binaries in one stage and copy the binaries to a second build stage, leaving the build artifacts from the first stage out of the final image. The new version of Docker should be out sometime this month.

**Comment by yb66** [comment]
@derikson I tried using multi-stage build today but I get `bash: /opt/rubinius/3.107/bin/rbx: No such file or directory` when running `rbx` (though it does exist) so I must've missed something.

```
FROM rubinius/docker as rubinius

# Localisation
# see https://github.com/Auswaschbar/alpine-localized-docker/blob/master/Dockerfile
FROM janu5/alpine-localised:3.9.2

ENV LC_ALL en_GB.UTF-8
ENV LANG en_GB.UTF-8
ENV LANGUAGE en_GB.UTF-8

# Added this bit in to mirror most of the packages that the
# Ubuntu version installs, in case they are run-time
# dependencies too.
RUN apk update && \
    apk upgrade && \
    apk add --no-cache dumb-init \
                       build-base \
                       libbz2 \
                       yaml-dev \
                       libressl \
                       clang \
                       libedit-dev \
                       zlib-dev \
                       bash \
                       make && \
    rm -rf /var/cache/apk/*

# Extract the install
RUN mkdir -p /opt/rubinius/3.107
COPY --from=rubinius /opt/rubinius/3.107 /opt/rubinius/3.107/



ENV PATH /opt/rubinius/3.107/bin:/opt/rubinius/3.107/gems/bin:$PATH
```

Regards,
iain

**Comment by CpuID** [comment]
To confirm there's an Ubuntu 16.04 Dockerfile in this repo, it has a few issues but it exists at least :)
